Prox/Prox/Data/PlacesController.swift

```diff
@@ -257,14 +270,20 @@ extension PlacesProvider: PlaceDataSource {
         }
     }
 
-    func allPlaces() -> [Place] {
-        return displayedPlaces
-    }
-
     func sortPlaces(byLocation location: CLLocation) {
         self.placesLock.withWriteLock {
             let sortedPlaces = PlaceUtilities.sort(places: displayedPlaces, byDistanceFromLocation: location)
             self.displayedPlaces = sortedPlaces
         }
     }
+
+    func refresh() {
+        assert(Thread.isMainThread)
+
+        placesLock.withWriteLock {
+            updateDisplayedPlaces()
+        }
+
+        delegate?.placesProvider(self, didUpdatePlaces: displayedPlaces)
```

Comment:
Yeah, I see we're doing this elsewhere in the file:
```
self.placesLock.withReadLock {
    self.delegate?.placesProvider(self, didUpdatePlaces: self.displayedPlaces)
}
```

That means we're holding the read lock for the entire execution of the `didUpdatePlaces` callback, so We're Gonna Have A Bad Time. Specifically [this](https://github.com/mozilla-mobile/prox/blob/eb86ad2227be304457fb526a041ab8ddb04af613/Prox/Prox/PlaceCarousel/PlaceCarouselViewController.swift#L449).